ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicaThread.java

```diff
@@ -76,6 +81,44 @@ public VcrReplicaThread(String threadName, FindTokenHelper findTokenHelper, Clus
     this.azureCloudConfig = new AzureCloudConfig(properties);
     this.azureTableNameReplicaTokens = this.azureCloudConfig.azureTableNameReplicaTokens;
     this.azureMetrics = new AzureMetrics(clusterMap.getMetricRegistry());
+    this.numReplIter = 0;
+  }
+
+  class ReplicaComparator implements Comparator<RemoteReplicaInfo> {
+    @Override
+    public int compare(RemoteReplicaInfo r1, RemoteReplicaInfo r2) {
+      String d1 = r1.getReplicaId().getDataNodeId().getHostname();
+      String d2 = r2.getReplicaId().getDataNodeId().getHostname();
+      return d1.compareTo(d2);
+    }
+  }
+
+  /**
+   * A custom filter that inheritors can override depending upon use case to pick which replicas to replicate from
+   * @param replicas A map of replicas {host -> {replicas}}
+   */
+  @Override
+  public Map<DataNodeId, List<RemoteReplicaInfo>> customFilter(Map<DataNodeId, List<RemoteReplicaInfo>> replicas) {
+    HashMap<Long, List<RemoteReplicaInfo>> partitions = new HashMap<>();
+    // Group replicas by partition
+    replicas.values().forEach(rlist -> rlist.forEach(replica -> {
+      long pid = replica.getReplicaId().getPartitionId().getId();
+      List alist = partitions.getOrDefault(pid, new ArrayList<>());
+      alist.add(replica);
+      partitions.putIfAbsent(pid, alist);
+    }));
+    // Pick one replica per partition in this iteration
+    Map<DataNodeId, List<RemoteReplicaInfo>> nodes = new HashMap<>();
+    partitions.values().forEach(rlist -> {
+      rlist.sort(new ReplicaComparator());
+      RemoteReplicaInfo replica = rlist.get(numReplIter % rlist.size());
+      DataNodeId dnode = replica.getReplicaId().getDataNodeId();
+      List alist = nodes.getOrDefault(dnode, new ArrayList<>());
+      alist.add(replica);
```

Comment:
`List` cannot be instantiated.

Reply:
I mean we can do something like this 
```
nodes.computeIfAbsent(dnode, k -> new ArrayList<>()).add(replica);
```
This can replace 
```
List alist = nodes.getOrDefault(dnode, new ArrayList<>());
alist.add(replica);
nodes.putIfAbsent(dnode, alist);
```

Reply:
Got it, i didn't know computeIfAbsent returns the entry value.